spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -174,41 +187,74 @@ protected Object build(Object payload, MessageHeaders headers) {
+		private final boolean input;
+
+		private final Class<?> klazz;
+
+		private final MessageWrappingMessageConverter messageConverter;
+
+		private ContentTypeConvertingInterceptor(String contentType, boolean input) {
 			this.contentType = contentType;
 			this.mimeType = MessageConverterUtils.getMimeType(contentType);
+			this.input = input;
+			if (MessageConverterUtils.X_JAVA_OBJECT.includes(this.mimeType)) {
+				this.klazz =
+						MessageConverterUtils
+								.getJavaTypeForJavaObjectContentType(this.mimeType);
+			}
+			else if (this.mimeType.equals(MessageConverterUtils.X_SPRING_TUPLE)) {
+				this.klazz = Tuple.class;
+			}
+			else if (this.mimeType.getType().equals("text") || this.mimeType.getSubtype().equals(
+					"json") || this.mimeType.getSubtype().equals("xml")) {
+				this.klazz = String.class;
+			}
+			else {
+				this.klazz = byte[].class;
+			}
+			this.messageConverter = new MessageWrappingMessageConverter(
+						MessageConverterConfigurer.this.compositeMessageConverterFactory
+								.getMessageConverterForType(this.mimeType), this.mimeType);
 		}
 
 		@Override
 		public Message<?> preSend(Message<?> message, MessageChannel channel) {
-			Class<?>[] classes =
-					MessageConverterConfigurer.this.compositeMessageConverterFactory.supportedDataTypes(
-							this.mimeType);
-			MessageWrappingMessageConverter messageConverter =
-					new MessageWrappingMessageConverter(
-							MessageConverterConfigurer.this.compositeMessageConverterFactory
-									.getMessageConverterForType(this.mimeType), this.mimeType);
-			for (Class<?> aClass : classes) {
-				if (aClass.isAssignableFrom(message.getPayload().getClass())) {
-					Object contentTypeFromMessage = message.getHeaders().get(MessageHeaders.CONTENT_TYPE);
-					if (contentTypeFromMessage == null) {
-						return MessageConverterConfigurer.this.messageBuilderFactory
-								.fromMessage(message)
-								.setHeader(MessageHeaders.CONTENT_TYPE, this.contentType)
-								.build();
-					}
-					else {
-						return message;
-					}
+			Message<?> sentMessage = null;
+			if (this.klazz.isAssignableFrom(message.getPayload().getClass())) {
+				Object contentTypeFromMessage = message.getHeaders().get(MessageHeaders.CONTENT_TYPE);
+				if (contentTypeFromMessage == null) {
+					sentMessage = MessageConverterConfigurer.this.messageBuilderFactory
+							.fromMessage(message)
+							.setHeader(MessageHeaders.CONTENT_TYPE, this.contentType)
+							.build();
 				}
 				else {
-					Object converted = messageConverter.fromMessage(message, aClass);
-					if (converted != null) {
-						return (Message<?>) converted;
+					sentMessage = message;
+				}
+			}
+			else {
+				Object converted = this.input ? this.messageConverter.fromMessage(message, this.klazz)
+						: this.messageConverter.toMessage(message.getPayload(), message.getHeaders(), null);
+				if (converted != null) {
+					Message<?> convertedMessage = (Message<?>) converted;
```

Comment:
No, the wrapper converter wraps it a `Message` of its own ... we could probably remove that and move the code here (the wrapper was required for DataType channels).